### PR TITLE
add a new IO named DataLakeIO (#23074)

### PR DESCRIPTION
**Pull request by zhangt-nhlab**
We developed  a new IO named DataLakeIO, which support beam to read data from data lake (delta, iceberg, hudi), and write data to data lake(delta, icberg, hudi).  

Because delta , iceberg and hudi does not provide enough java api to read and write, so we use spark datasouce api to read and write data in DataLakeIO. Therefore, the spark dependencies is needed.   

BeamDeltaTest, BeamIcebergTest and BeamHudiTest show how to use the above features.

### CODE REVIEW

**Comment by zhangt-nhlab** [comment]
@kileys 
@Abacn
@johnjcasey

File: sdks/java/io/datalake/src/main/java/org/apache/beam/sdk/io/datalake/DataLakeIO.java [hunk 1]
```diff
@@ -0,0 +1,530 @@
...
+                builder.add(DisplayData.item("query", getQuery()));
+            }
+            builder.add(DisplayData.item("rowMapper", getRowMapper().getClass().getName()));
+            if (getCoder() != null) {
+                builder.add(DisplayData.item("coder", getCoder().getClass().getName()));
+            }
+        }
+    }
+
+    private static class ReadFn<ParameterT, OutputT> extends DoFn<ParameterT, OutputT> {
```
Comment (johnjcasey) [inline sdks/java/io/datalake/src/main/java/org/apache/beam/sdk/io/datalake/DataLakeIO.java:374]: Currently, we are trying to have all sources implemented using the SplittableDoFn pattern to enable scalability, and are doing our best to not include new sources that are not implemented as SDFs. Can this be re-implemented as an SDF instead?

File: sdks/java/io/datalake/build.gradle [hunk 2]
```diff
@@ -0,0 +1,54 @@
+/*
```
Comment (johnjcasey) [inline sdks/java/io/datalake/build.gradle:1]: Can you re-work these dependencies to match the pattern used for other IOs? See io/google-cloud-platform/build.gradel for an example.

New dependencies themselves are included buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy

**Review by johnjcasey — CHANGES_REQUESTED** [review]
As a general comment, it may also be worth looking at the in-progress SparkReceiverIO to see if a more generic spark connection can be used.

**Comment by Abacn** [comment]
waiting on author

File: sdks/java/io/datalake/build.gradle [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+description = "Apache Beam :: SDKs :: Java :: IO :: Datalake"
+ext.summary = "IO to read from and write to Data Lake"
+
+dependencies {
+    implementation project(path: ":sdks:java:core", configuration: "shadow")
+//    implementation library.java.slf4j_api
+    implementation library.java.vendored_guava_26_0_jre
+    implementation "org.apache.spark:spark-sql_2.12:3.1.2"
+    implementation "org.apache.spark:spark-core_2.12:3.1.2"
+    implementation "org.apache.spark:spark-streaming_2.12:3.1.2"
```
Comment (mosche) [inline sdks/java/io/datalake/build.gradle:33]: This IO would be really neat and I understand the motivation of using Spark underneath.
Nevertheless, the spark dependency is rather problematic and I'm very concerned about the consequences ... 

There's also a Spark runner, which supports both Spark 2.4 and Spark >= 3.1. This IO would certainly conflict with the Spark 2.4 runner. The Spark 3 runner is build in a way that it supports various versions of Spark 3 (the path from 3.1 to 3.3 is full of breaking changes), Spark dependencies are typically provided (as available on the cluster). Even further, Spark comes with a massive tail of dependencies prone to causing conflicts with versions used in Beam.

The one common candidate to mention here is Avro. Spark 3.1 is still using Avro 1.8 matching Beam's version, Spark 3.2 bumps Avro to 1.10 which is incompatible with Beam :/ This kinda exemplifies the maintenance headache ahead.

Have you evaluated any alternative to using Spark underneath?

Comment (mosche) [reply sdks/java/io/datalake/build.gradle:33]: cc @aromanenko-dev

**Review by aromanenko-dev — CHANGES_REQUESTED** [review]
First of all - thanks for your contribution!

Before proceeding to review from my side, I'd like to know if there is a design doc or similar for this IO connector? It would be very helpful to understand the goals and the implementation of this connector in advance. 

Also, several notes that are worth to mention:
1) Please, create a new github issue for this feature.
2) Please, avoid merging a `master` branch into your feature branch. Use `git rebase` instead.
3) Run `./gradlew :sdks:java:io:datalake:check` locally before pushing your changes to origin.

You can find a Beam contribution guide here: 
https://beam.apache.org/contribute/get-started-contributing/

**Comment by zhangt-nhlab** [comment]
> First of all - thanks for your contribution!
> 
> Before proceeding to review from my side, I'd like to know if there is a design doc or similar for this IO connector? It would be very helpful to understand the goals and the implementation of this connector in advance.
> 
> Also, several notes that are worth to mention:
> 
> 1. Please, create a new github issue for this feature.
> 2. Please, avoid merging a `master` branch into your feature branch. Use `git rebase` instead.
> 3. Run `./gradlew :sdks:java:io:datalake:check` locally before pushing your changes to origin.
> 
> You can find a Beam contribution guide here: https://beam.apache.org/contribute/get-started-contributing/

**Comment by zhangt-nhlab** [comment]
> > First of all - thanks for your contribution!
> > Before proceeding to review from my side, I'd like to know if there is a design doc or similar for this IO connector? It would be very helpful to understand the goals and the implementation of this connector in advance.
> > Also, several notes that are worth to mention:
> > 
> > 1. Please, create a new github issue for this feature.
> > 2. Please, avoid merging a `master` branch into your feature branch. Use `git rebase` instead.
> > 3. Run `./gradlew :sdks:java:io:datalake:check` locally before pushing your changes to origin.
> > 
> > You can find a Beam contribution guide here: https://beam.apache.org/contribute/get-started-contributing/

Thank you for your reply! I will make my changes, and create a new github issue later.

File: sdks/java/io/datalake/build.gradle (hunk 3 above)
Comment (zhangt-nhlab) [reply sdks/java/io/datalake/build.gradle:33]: > This IO would be really neat and I understand the motivation of using Spark underneath. Nevertheless, the spark dependency is rather problematic and I'm very concerned about the consequences ...
> 
> There's also a Spark runner, which supports both Spark 2.4 and Spark >= 3.1. This IO would certainly conflict with the Spark 2.4 runner. The Spark 3 runner is build in a way that it supports various versions of Spark 3 (the path from 3.1 to 3.3 is full of breaking changes), Spark dependencies are typically provided (as available on the cluster). Even further, Spark comes with a massive tail of dependencies prone to causing conflicts with versions used in Beam.
> 
> The one common candidate to mention here is Avro. Spark 3.1 is still using Avro 1.8 matching Beam's version, Spark 3.2 bumps Avro to 1.10 which is incompatible with Beam :/ This kinda exemplifies the maintenance headache ahead.
> 
> Have you evaluated any alternative to using Spark underneath?

Your advice is great !  I'll consider it, and then think about other alternatives

File: sdks/java/io/datalake/build.gradle (hunk 2 above)
Comment (zhangt-nhlab) [reply sdks/java/io/datalake/build.gradle:1]: Okay, I'll modify it according to this

File: sdks/java/io/datalake/src/main/java/org/apache/beam/sdk/io/datalake/DataLakeIO.java (hunk 1 above)
Comment (zhangt-nhlab) [reply sdks/java/io/datalake/src/main/java/org/apache/beam/sdk/io/datalake/DataLakeIO.java:374]: Thanks, I'll learn SplittableDoFn, then re-implemente it  as an SDF instead.

**Comment by aaltay** [comment]
Was there any progress on getting this IO into Beam?

**Pull request closed without merging**